### PR TITLE
[clang] Provide an SSE4.2 implementation of identifier token lexer

### PR DESCRIPTION
**Pull request by serge-sans-paille**
The _mm_cmpistri instruction can be used to quickly parse identifiers.

With this patch activated, clang pre-processes <iostream> 1.8% faster, and sqlite3.c amalgametion 1.5% faster, based on time measurements and number of executed instructions as measured by valgrind.

The introduction of an extra helper function in the regular case has no impact on performance, see

    https://llvm-compile-time-tracker.com/compare.php?from=30240e428f0ec7d4a6d1b84f9f807ce12b46cfd1&to=12bcb016cde4579ca7b75397762098c03eb4f264&stat=instructions:u

### CODE REVIEW

**Review by cor3ntin — COMMENTED** [review]
This is pretty neat.
I'm wondering if this change would be tested by any of our bots?

File: clang/lib/Lex/Lexer.cpp [hunk 1]
```diff
@@ -1847,19 +1851,46 @@ bool Lexer::LexUnicodeIdentifierStart(Token &Result, uint32_t C,
...
 
+static const char *fastParseASCIIIdentifier(const char *CurPtr, const char* BufferEnd) {
+#ifdef __SSE4_2__
+  static constexpr char AsciiIdentifierRange[16] = {
+      '_', '_', 'A', 'Z', 'a', 'z', '0', '9',
+  };
+  constexpr ssize_t BytesPerRegister = 16;
+
+  while (LLVM_LIKELY(BufferEnd - CurPtr >= BytesPerRegister)) {
+    __m128i AsciiIdentifierRangeV = _mm_loadu_si128((const __m128i *)AsciiIdentifierRange);
```
Comment (tbaederr) [inline clang/lib/Lex/Lexer.cpp:1862]: Can't you pull this out of the loop?

And can't you do an aligned load here (if you add the right `alignas()` to the declaration)?

Comment (serge-sans-paille) [reply clang/lib/Lex/Lexer.cpp]: LICM + alignment done!

**Comment by serge-sans-paille** [comment]
@tbaederr / @cor3ntin Do I have your formal approval to merge this?

**Comment by cor3ntin** [comment]
@serge-sans-paille I'm still trying to understand whether this change will be tested by CI - or at all.
SSE4 is not enabled by default and i don't think our cmake does that either - I did ping @AaronBallman and we are still thinking about it

**Comment by serge-sans-paille** [comment]
ok. I obviously tested on my setup, but having consistent testing is important. Thanks for the quick answer!

**Comment by AaronBallman** [comment]
> @serge-sans-paille I'm still trying to understand whether this change will be tested by CI - or at all. SSE4 is not enabled by default and i don't think our cmake does that either - I did ping @AaronBallman and we are still thinking about it

I think we're in good shape for testing because we have at least one bot that's set up to compile for cascade lake, which does support SSE 4.2: https://github.com/llvm/llvm-zorg/blob/4bc450165fa9fcef26adca5a80f4d97b9d6babc3/buildbot/osuosl/master/config/builders.py#L827 so I believe that gets us the test coverage we need. Any disagreement?

**Review by tbaederr — APPROVED** [review]
LGTM then

**Pull request merged**